File: src/bfs_dfs/TravelPath.java
package bfs_dfs;

import java.util.*;

//https://programmers.co.kr/learn/courses/30/lessons/43164
public class TravelPath {
    private List<String> paths = new LinkedList<>();

    private List<Integer> getAvailable(boolean[] visit, String departure, String[][] tickets) {
        List<Integer> available = new LinkedList<>();
        for (int i = 0; i < tickets.length; i++) {
            if (departure.equals(tickets[i][0]) && !visit[i])
                available.add(i);
        }
        return available;
    }

    private boolean isAllVisited(boolean[] visit) {
        for (boolean b : visit) {
            if (!b) return false;
        }
        return true;
    }

    private void dfs(boolean[] visit, List<String> path, String[][] tickets, String departure) {
        List<Integer> available = getAvailable(visit, departure, tickets);
        path.add(departure + " ");
        for (Integer a : available) {
            boolean[] v = visit.clone();
            List<String> p = new LinkedList<>(path);
            if (isAllVisited(v))
                return;
            v[a] = true;
            dfs(v, p, tickets, tickets[a][1]);
        }

        if (path.size() == tickets.length + 1)
            paths.add(String.join("", path));
    }

    public String[] solution(String[][] tickets) {
        boolean[] visit = new boolean[tickets.length];
        List<String> path = new LinkedList<>();
        dfs(visit, path, tickets, "ICN");
        paths.sort(String::compareTo);
        System.out.println(paths);
        return paths.get(0).split(" ");
    }

    public static void main(String[] args) {
        String[][] tickets =
                {{"ICN", "SFO"},
                        {"ICN", "ATL"},
                        {"SFO", "ATL"},
                        {"ATL", "ICN"},
                        {"ATL", "SFO"}};
        System.out.println(Arrays.toString(new TravelPath().solution(tickets)));
    }
}

File: src/bfs_dfs/travel_path.py
# 여행 경로
# https://programmers.co.kr/learn/courses/30/lessons/43164
from collections import defaultdict,deque
import copy
def solution(tickets):
    nextpaths = defaultdict(list)
    cand = []
    n = len(tickets)
    for tidx, ticket in enumerate(tickets):
        a,b=ticket
        nextpaths[a].append((b, tidx))
    q = deque([["", "ICN",-1, 0, [], defaultdict(bool)]])
    while q:
        start,dest,tidx,usecount, path, visit = q.popleft()
        path.append(dest)
        visit[tidx]=True
        if usecount == n:
            cand.append(path)
            continue
        usecount+=1
        start = dest
        for dest,tidx in nextpaths[start]:
            if not visit[tidx]:
                q.append([start, dest, tidx, usecount, copy.copy(path), copy.copy(visit)])            
    return min(cand)

print(solution([["ICN", "SFO"], ["ICN", "ATL"], ["SFO", "ATL"], ["ATL", "ICN"], ["ATL", "SFO"]]))
# ["ICN", "ATL", "ICN", "SFO", "ATL", "SFO"]